### DISCUSSION TITLE
Deterministic Replacements

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#4537] [Language Ideas] [answered]
If I have a deterministic function/expression that is called with constant parameters, it would be nice if C# would evaluate the expression at compile time and replace the expression with the value.

**Comment by Joe4evr** [comment, marked as answer]
#2379

**Comment by Unknown6656** [comment]
I'll just mention the other related issues/proposals here:

- [dotnet/csharplang#1763](https://github.com/dotnet/csharplang/issues/1763): "[C#-X.0 Proposal] Determinism"
- [dotnet/csharplang#504](https://github.com/dotnet/csharplang/issues/504): "Enable Compile-Time Folding of Deterministic Outputs"
- [dotnet/csharplang#776](https://github.com/dotnet/csharplang/issues/776): "C# Pure Function Keyword to Mark No Side Effects Or External Dependencies" 
- [dotnet/csharplang#1028](https://github.com/dotnet/csharplang/issues/1028): "Compile time expressions"
- [dotnet/csharplang#1413](https://github.com/dotnet/csharplang/issues/1413): "Compile-Time method inlining"
- [dotnet/roslyn#9627](https://github.com/dotnet/roslyn/issues/9627): "Compile Time Function Execution"
- [dotnet/roslyn#10506](https://github.com/dotnet/roslyn/issues/10506): "Static Expressions"
- [dotnet/roslyn#11259](https://github.com/dotnet/roslyn/issues/11259): "Constant string interpolation"
- [dotnet/roslyn#12238](https://github.com/dotnet/roslyn/issues/12238): "Compile time expressions"
- [dotnet/roslyn#14665](https://github.com/dotnet/roslyn/issues/14665): "Constant expression - Power operator?"
- [dotnet/roslyn#15079](https://github.com/dotnet/roslyn/issues/15079): "How would you imagine `constexpr` in C#?"
- [dotnet/coreclr#3633](https://github.com/dotnet/coreclr/issues/3633): "Optimizing `constant_string.Length`"
